### PR TITLE
ffmpeg-qsv/ffmpeg-vaapi: enable hwmap testing

### PR DESCRIPTION
**Pull request by yefengxx**
add hwmap testing implement for ffmpeg-qsv and ffmpeg-vaapi

Signed-off-by: Xu Yefeng <yefengx.xu@intel.com>

### CODE REVIEW

**Comment by yefengxx** [comment]
@uartie @FocusLuo  can you review this for me?

**Comment by uartie** [comment]
Also, are there some technical docs about this feature?

File: test/ffmpeg-vaapi/vpp/hwmap.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+###
+
+from ....lib import *
+from ....lib.ffmpeg.vaapi.util import *
+from ....lib.ffmpeg.vaapi.vpp import VppTest
+
+spec      = load_test_spec("vpp", "vaapi_hwmap")
+
+@slash.requires(*platform.have_caps("vpp", "scale"))
+@slash.requires(*have_ffmpeg_filter("scale_qsv"))
```
Comment (uartie) [inline test/ffmpeg-vaapi/vpp/hwmap.py:14]: What if the test case isn't configured to use the "scale_qsv" vpp_op?

**Comment by uartie** [comment]
I don't see all necessary checks for the requirements.  For example, what if scale_vulkan is not built/installed on target device?  It should skip the test case if that happens.

After a further quick look of this change, these feel more like transcode tests with vpp scaling.... just mixing different hw-based plugins to do the vpp and encode parts.  I wonder if it makes more sense to add this support to transcode tests, instead?

**Comment by uartie** [comment]
In a bigger picture, this is more about mixing/interop between various decode, vpp, encode plugins.  The "hwmap" term is really only a ffmpeg concept and the means (syntax) to achieve interop between different hw-based codec elements in ffmpeg.

Furthermore, "raw" coding can also be categorized as a special codec for decode/encode in uncompressed/basic format (which we exploit a lot in this framework).  It just does not require extra "hwmap" syntax/concepts.  And the logic to handle this type of input/output is mostly encapsulated "under-the-hood". 

And what about gstreamer?  I'm sure we want to start testing mixed/interop elements here, too.  And it's likely more about caps/memory syntax in gstreamer rather than "hwmap" syntax.

Thus, I think "hwmap" (i.e. the logic to hook from one hw-element to another) should be totally encapsulated, too, since it can be defined differently between various middleware.

So, we really need to think about how we can **generalize** this concept of mixing elements/plugins without exposing middleware-specific terms into the test framework API's and especially into the higher-level user test-case configurations.

**Comment by yefengxx** [comment]
> I don't see all necessary checks for the requirements. For example, what if scale_vulkan is not built/installed on target device? It should skip the test case if that happens.
> 
> After a further quick look of this change, these feel more like transcode tests with vpp scaling.... just mixing different hw-based plugins to do the vpp and encode parts. I wonder if it makes more sense to add this support to transcode tests, instead?

This test is for hwmap. The VPP scaling and encoding operations are just to verify whether the hwmap operation is successful. If we don't scale and encode, we won't know if hwmap is successful.

**Comment by uartie** [comment]
Needs more thorough test design plan.  Close.

**Pull request closed without merging**